plugins/inputs/kafka_consumer/kafka_consumer.go

```diff
@@ -143,9 +151,138 @@ func (k *KafkaConsumer) Init() error {
+
+	k.compileTopicRegexps()
+
+	if len(k.regexps) == 0 {
+		// There are no regexp-matched topics
+		k.allWantedTopics = k.Topics
+	}
+
 	return nil
 }
 
+func (k *KafkaConsumer) createTopicScanner() error {
+	if k.topicClient == nil {
+		client, err := sarama.NewClient(k.Brokers, k.config)
+		if err != nil {
+			return err
+		}
+		k.topicClient = client
+	}
+	return nil
+}
+
+func (k *KafkaConsumer) compileTopicRegexps() {
+	// While we can add new topics matching extant regexps, we can't
+	// update that list on the fly.  We compile them once at startup.
+	// Changing them is a configuration change and requires a restart.
+
+	if len(k.TopicRegexps) == 0 {
+		// Nothing to do.
+		return
+	}
+	k.regexps = make([]regexp.Regexp, 0, len(k.TopicRegexps))
+	// Let's optimistically assume that all of them compile
+	for _, r := range k.TopicRegexps {
+		re, err := regexp.Compile(r)
+		if err == nil {
+			k.regexps = append(k.regexps, *re)
+			k.Log.Infof("Added regular expression '%s' to topics", r)
+		} else {
+			k.Log.Errorf("Regular expression '%s' did not compile: '%w'; not adding to topics", r, err)
+		}
+	}
+}
+
+func (k *KafkaConsumer) refreshTopics() error {
+	// We have instantiated a new generic Kafka client, so we can ask
+	// it for all the topics it knows about.  Then we build
+	// regexps from our strings, loop over those, loop over the
+	// topics, and if we find a match, add that topic to
+	// out topic set, which then we turn back into a list at the end.
+
+	if len(k.regexps) == 0 {
+		return nil
+	}
+	if k.topicClient == nil {
+		err := k.createTopicScanner()
+		if err != nil {
+			return fmt.Errorf("could not create client to match topics to regexps: %w", err)
+		}
+	}
+
+	allDiscoveredTopics, err := k.topicClient.Topics()
+	if err != nil {
+		return err
+	}
+	k.Log.Infof("discovered topics: %v", allDiscoveredTopics)
+
+	extantTopicSet := make(map[string]bool, len(allDiscoveredTopics))
+	for _, t := range allDiscoveredTopics {
+		extantTopicSet[t] = true
+	}
+	// Even if a topic specified by a literal string (that is, k.Topics)
+	// does not appear in the topic list, we want to keep it around, in
+	// case it pops back up--it is not guaranteed to be matched by any
+	// of our regular expressions.  Therefore, we pretend that it's in
+	// extantTopicSet, even if it isn't.
+	//
+	// Assuming that literally-specified topics are usually in the topics
+	// present on the broker, this should not need a resizing (although if
+	// you have many topics that you don't care about, it will be too big)
+	wantedTopicSet := make(map[string]bool, len(allDiscoveredTopics))
+	for _, t := range k.Topics {
+		// Get our pre-specified topics
+		k.Log.Infof("adding literally-specified topic %s", t)
+		wantedTopicSet[t] = true
+	}
+	for t := range extantTopicSet {
+		// Add topics that match regexps
+		for _, r := range k.regexps {
+			if r.MatchString(t) {
+				wantedTopicSet[t] = true
+				k.Log.Infof("adding regexp-matched topic '%s'", t)
```

Comment:
Please be careful with the logging levels as this function is called repeatedly these messages might spam the log-file. Use the `debug` functions for this kind of messages.